src/core/provider/MCAccountValidator.cpp

```diff
@@ -0,0 +1,421 @@
+            
+            mOperation = (IMAPOperation *)mImapSession->checkAccountOperation();
+            mOperation->setCallback((OperationCallback *)this);
+            mOperation->start();
+        
+        }
+        else {
+            mCurrentServiceTested ++;
+            mCurrentServiceIndex = 0;
+            checkNextHost();
+        }
+    }
+    else if (mCurrentServiceTested == SERVICE_POP){
+        if (mCurrentServiceIndex < mPopServices->count()) {
+            POPAsyncSession *popSession = new POPAsyncSession();
+            popSession->setUsername(mUsername);
+            popSession->setPassword(mPassword);
+            
+            mPopServer = (NetService *) mPopServices->objectAtIndex(mCurrentServiceIndex);
+            popSession->setHostname(mPopServer->hostname());
+            popSession->setPort(mPopServer->port());
+            popSession->setConnectionType(mPopServer->connectionType());
+            
+            mOperation = (POPOperation *)popSession->checkAccountOperation();
+            mOperation->setCallback((OperationCallback *)this);
+            mOperation->start();
+        }
+        else {
+            mCurrentServiceTested ++;
+            mCurrentServiceIndex = 0;
+            checkNextHost();
+        }
+    }
+    else if (mCurrentServiceTested == SERVICE_SMTP){
+        if (mCurrentServiceIndex < mSmtpServices->count()) {
+            SMTPAsyncSession *smtpSession = new SMTPAsyncSession();
+            smtpSession->setUsername(mUsername);
+            smtpSession->setPassword(mPassword);
+            
+            mSmtpServer = (NetService *) mSmtpServices->objectAtIndex(mCurrentServiceIndex);
+            smtpSession->setHostname(mSmtpServer->hostname());
+            smtpSession->setPort(mSmtpServer->port());
+            smtpSession->setConnectionType(mSmtpServer->connectionType());
+            
+            mOperation =  (SMTPOperation *)smtpSession->checkAccountOperation(Address::addressWithMailbox(mEmail));
+            mOperation->setCallback((OperationCallback *)this);
+            mOperation->start();
+
+        }
+        else {
+            mCurrentServiceTested ++;
+            mCurrentServiceIndex = 0;
+            checkNextHost();
+        }
+    }
+    else {
+        MC_SAFE_RELEASE(mPopSession);
+        MC_SAFE_RELEASE(mSmtpSession);
+        callback()->operationFinished(this);
+    }
+}
+
+void AccountValidator::checkNextHostDone()
+{
+    ErrorCode error = ErrorNone;
+    
+    if (mCurrentServiceTested == SERVICE_IMAP) {
+        mImapError = ((IMAPOperation *)mOperation)->error();
+        error = mImapError;
+    }
+    else if (mCurrentServiceTested == SERVICE_POP) {
+        mPopError = ((POPOperation *)mOperation)->error();
+        error = mPopError;
+    }
+    else if (mCurrentServiceTested == SERVICE_SMTP) {
+        mSmtpError = ((SMTPOperation *)mOperation)->error();
+        error = mSmtpError;
+    }
+    
```

Comment:
If I release each session here after each use. 
This means I do not need to do it at the of checkNextHost() just before callback()->operationFinished(this); right?